support/cas-server-support-duo-core/src/main/java/org/apereo/cas/adaptors/duo/authn/DuoAuthenticationHandler.java

```diff
@@ -102,26 +97,15 @@ private AuthenticationHandlerExecutionResult authenticateDuoCredential(final Cre
         }
     }
 
-    private DuoSecurityAuthenticationService getDuoAuthenticationService() {
-        final RequestContext requestContext = RequestContextHolder.getRequestContext();
-        if (requestContext == null) {
-            throw new IllegalArgumentException("No request context is held to locate the Duo authentication service");
-        }
-        final Collection<String> providerIds = WebUtils.getResolvedMultifactorAuthenticationProviders(requestContext);
-        final Collection<MultifactorAuthenticationProvider> providers =
-            MultifactorAuthenticationUtils.getMultifactorAuthenticationProvidersByIds(providerIds,
-                ApplicationContextProvider.getApplicationContext());
-
-        if (providers.isEmpty()) {
-            throw new IllegalArgumentException("No multifactor providers are found in the current request context");
-        }
-        final MultifactorAuthenticationProvider pr = providers.iterator().next();
-        return provider.findProvider(pr.getId(), DuoMultifactorAuthenticationProvider.class).getDuoAuthenticationService();
+    @Override
+    public String getProviderId() {
+        return provider.getId();
     }
 
     @Override
     public boolean supports(final Credential credential) {
-        return DuoCredential.class.isAssignableFrom(credential.getClass())
-            || credential instanceof DuoDirectCredential;
+        return (DuoCredential.class.isAssignableFrom(credential.getClass())
+                || credential instanceof DuoDirectCredential)
+                && supports((MfaCredential) credential);
```

Comment:
Is `supports((MfaCredential) credential);` a recursive call that would cause an overflow issue?

Reply:
Not that I found in my testing.  It calls the overloaded default method in MfaAuthenticationHandler interface.